backend/de.metas.swat/de.metas.swat.base/src/main/java/org/adempiere/inout/util/ShipmentScheduleQtyOnHandStorageHolder.java

```diff
@@ -0,0 +1,40 @@
+package org.adempiere.inout.util;
+
+import com.google.common.collect.ImmutableList;
+import de.metas.inoutcandidate.api.OlAndSched;
+import de.metas.inoutcandidate.model.I_M_ShipmentSchedule;
+import de.metas.material.cockpit.stock.StockDataQuery;
+import lombok.AllArgsConstructor;
+import lombok.NonNull;
+
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+
+/**
+ * Loads stock details which are relevant to given {@link I_M_ShipmentSchedule}s.
+ * Allows to change (in memory!) the qtyOnHand.
+ */
+@AllArgsConstructor(staticName="of")
+public class ShipmentScheduleQtyOnHandStorageHolder
+{
+	private final ImmutableList<IShipmentScheduleQtyOnHandStorage> storages;
+
+	@NonNull
+	public Optional<StockDataQuery> toQuery(@NonNull final I_M_ShipmentSchedule sched)
+	{
+		return storages.stream()
+				.map(storage -> storage.toQuery(sched))
+				.filter(Objects::nonNull)
+				.findFirst();
```

Comment:
not ok from my POV, I'd go with the full list

Reply:
This is used just as a stupid logging statement, however because caches are in `...Storage`, and the `StockDataQuery` only makes sense in the context of QtyOnHand, I chose to design it like this.
The alternative would have been to make it a static method on `ShipmentScheduleQtyOnHandStorage` and either pass in the instance (to have access to the caches) or ditch the caching altogether and in the static always compute the value from scratch. I chose the all-but-ShipmentScheduleQtyOnHandStorage-return-null approach but am open to other suggrestions.

Reply:
i also have some doubts about findFirst.... let's talk tomorrow

Reply:
As discussed, no better solution can be found within budget.